Sources/CvPlayerAI.cpp

```diff
@@ -12852,9 +12852,58 @@ int CvPlayerAI::AI_unitPropertyValue(UnitTypes eUnit, PropertyTypes eProperty) c
 	return iValue;
 }
 
-/********************************************************************************/
-/* 	City Defenders						24.07.2010				Fuyu			*/
-/********************************************************************************/
+int CvPlayerAI::AI_unitBuildingValue(UnitTypes eUnit, CvArea* pArea) const
+{
+	FAssertMsg(eUnit != NO_UNIT, "Unit is not assigned a valid value");
+	FAssertMsg(eUnitAI != NO_UNITAI, "UnitAI is not assigned a valid value");
+
+	CvUnitInfo& kUnitInfo = GC.getUnitInfo(eUnit);
+
+	// Evaluate unit's ability to construct buildings
+	int iCount = 0;
+
+	if (kUnitInfo.getNumBuildings() > 0 && pArea != NULL && !isNPC())
+	{
+		int iBuildingValue;
+		int iLoop;
+		bool bCoastal = kUnitInfo.getDomainType() == DOMAIN_SEA;
+		int iMinOceanSize;
+		if (bCoastal)
+		{
+			iMinOceanSize = GC.getMIN_WATER_SIZE_FOR_OCEAN();
+		}
+
+		for (int iI = 0; iI < kUnitInfo.getNumBuildings(); iI++)
+		{
+			BuildingTypes eBuilding = (BuildingTypes)kUnitInfo.getBuildings(iI);
+
+			if (NO_BUILDING != eBuilding)
+			{
+				if (canConstruct(eBuilding, false, false, true) && AI_getNumBuildingsNeeded(eBuilding, bCoastal) > 0)
+				{
+					iCount = 0;
+					iBuildingValue = 0;
+					for (CvCity* pLoopCity = firstCity(&iLoop); pLoopCity != NULL; pLoopCity = nextCity(&iLoop))
```

Comment:
I bet looping over city is slower than looping on the outer loop. You could instead do:
```
foreach city
    if city is costal etc
        for each unit building { ... }
```
i.e. Invert the loop so that you only do the slow loop once.

Reply:
Hah... makes sense and a great observation.  I'll look for that in the future.  My thinking was why loop the cities if you don't have to but you WILL have to at least once so I'm setting us up to loop through them more than that rather than to avoid having to at all.  Right.